tests/migrations.rs

```diff
@@ -114,3 +115,69 @@ fn check_non_null_column_without_default() {
         }
     }
 }
+
+fn get_up_migrations() -> Vec<PathBuf> {
+    let root = env!("CARGO_MANIFEST_DIR");
+    let migrations_dir = PathBuf::from(root).join("migrations");
+
+    let mut migrations = Vec::new();
+    for entry in std::fs::read_dir(migrations_dir).unwrap() {
+        let path = entry.unwrap().path();
+
+        if path.is_file() && path.to_string_lossy().contains(".up.sql") {
+            migrations.push(path);
+        }
+    }
+
+    // Sort by timestamp to enforce migration order
+    migrations.sort_by(|a, b| a.file_name().unwrap().cmp(b.file_name().unwrap()));
+    migrations
+}
+
+fn get_test_data_path(migration_path: &Path) -> PathBuf {
+    let filename = migration_path.file_name().unwrap().to_str().unwrap();
+    // e.g. "20240517094752_create_build.up.sql" -> "20240517094752_create_build"
+    let migration_name = filename.split('.').next().unwrap();
+
+    let root = env!("CARGO_MANIFEST_DIR");
+    PathBuf::from(root)
+        .join("tests/data/migrations")
+        .join(format!("{}.sql", migration_name))
+}
+
+#[test]
+fn check_migrations_have_sample_data() {
+    for migration_path in get_up_migrations() {
+        let test_data_path = get_test_data_path(&migration_path);
+
+        assert!(
+            test_data_path.exists(),
+            "Migration {:?} does not have an associated test data file at {:?}",
+            migration_path,
+            test_data_path
+        );
+    }
+}
+
+#[sqlx::test(migrations = false)]
+async fn apply_migrations_with_test_data(pool: PgPool) -> sqlx::Result<()> {
+    let migrations = get_up_migrations();
+
+    for migration_path in migrations {
+        let migration_sql = std::fs::read_to_string(&migration_path)
+            .unwrap_or_else(|_| panic!("Failed to read migration file: {:?}", migration_path));
+
+        pool.execute(&*migration_sql)
+            .await
+            .unwrap_or_else(|e| panic!("Failed to apply migration {:?}: {}", migration_path, e));
+
+        let test_data_path = get_test_data_path(&migration_path);
+        if let Ok(test_data) = std::fs::read_to_string(&test_data_path) {
```

Comment:
I know that we have the separate test for checking that this file exists, but we should still panic here if it doesn't. Otherwise the test could e.g. skip one migration and then report an unrelated error, which could be confusing. So let's just `expect()` here.